src/js/DOMUtils.js

```diff
@@ -0,0 +1,197 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+goog.provide('axs.dom');
+
+/**
+ * Returns the nearest ancestor which is an Element.
+ * @param {Node} node
+ * @return {?Element}
+ */
+axs.dom.parentElement = function(node) {
+    if (!node)
+        return null;
+
+    var parentNode = axs.dom.composedParentNode(node);
+    if (!parentNode)
+        return null;
+
+    switch (parentNode.nodeType) {
+    case Node.ELEMENT_NODE:
+        return /** @type {Element} */ (parentNode);
+    default:
+        return axs.dom.parentElement(parentNode);
+    }
+};
+
+
+(function() {
+
+/**
+ * Generate a random ID using crypto.getRandomValues().
+ * Taken from http://stackoverflow.com/a/27747377
+ * @return {string}
+ */
+function generateId() {
+    var len = 40;
+    var arr = new Uint8Array(len / 2);
+    window.crypto.getRandomValues(arr);
+    return [].map.call(arr, function(n) { return n.toString(16); }).join("");
+}
+
+/**
+ * Returns the given Node's parent in the composed tree.
+ * @param {Node} node
+ * @return {?Node}
+ */
+axs.dom.composedParentNode = function(node) {
+    if (!node)
+        return null;
+    if (node.nodeType === Node.DOCUMENT_FRAGMENT_NODE)
+        return node.host;
+
+    var parentNode = node.parentNode;
+    if (!parentNode)
+        return null;
+
+    if (parentNode.nodeType === Node.DOCUMENT_FRAGMENT_NODE)
+        return parentNode.host;
+
+    var shadowRoot = parentNode.shadowRoot;
+    if (!parentNode.shadowRoot)
+        return parentNode;
+
+    var eventName = generateId();
+    var path;
+    parentNode.addEventListener(eventName, function(event) {
+       event.stopPropagation();
+       path = event.path;
+       return false;
+    })
+    var event = new CustomEvent(eventName, { bubbles: true });
+    node.dispatchEvent(event);
+    for (var i = 1; i < path.length; i++) {
+        var pathNode = path[i];
+        if (pathNode.localName == 'content')
+            continue;
+        if (pathNode.nodeType === Node.DOCUMENT_FRAGMENT_NODE)
+            continue;
+        return pathNode;
+    }
+    return null;
+};
+})();
+
+/**
+ * Return the corresponding element for the given node.
+ * @param {Node} node
+ * @return {Element}
+ * @suppress {checkTypes}
+ */
+axs.dom.asElement = function(node) {
+    /** @type {Element} */ var element;
+    switch (node.nodeType) {
+    case Node.COMMENT_NODE:
+        return null;  // Skip comments
+    case Node.ELEMENT_NODE:
+        element = /** (@type {Element}) */ node;
+        if (element.localName == 'script')
+            return null;  // Skip script elements
+        break;
+    case Node.TEXT_NODE:
+        element = axs.dom.parentElement(node);
+        break;
+    default:
+        console.warn('Unhandled node type: ', node.nodeType);
```

Comment:
Actually, not so much the latter, but it still makes sense to handle DOCUMENT_FRAGMENT_NODE here.